qiskit_machine_learning/primitives/kernels/base_kernel.py

```diff
@@ -0,0 +1,117 @@
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Base kernel"""
+
+from abc import abstractmethod
+from typing import Tuple
+
+import numpy as np
+
+
+class BaseKernel:
+    """
+    Abstract class providing the interface for the quantum kernel classes.
+    """
+
+    def __init__(self, *, enforce_psd: bool = True) -> None:
+        """
+        Args:
+            sampler_factory: A callable that creates a qiskit primitives sampler
+                given a list of circuits.
+            enforce_psd: Project to closest positive semidefinite matrix if x = y.
+                Default True.
+        """
+        self._num_features: int = 0
+        self._enforce_psd = enforce_psd
+
+    @abstractmethod
+    def evaluate(self, x_vec: np.ndarray, y_vec: np.ndarray = None) -> np.ndarray:
+        r"""
+        Construct kernel matrix for given data
+
+        If y_vec is None, self inner product is calculated.
+
+        Args:
+            x_vec: 1D or 2D array of datapoints, NxD, where N is the number of datapoints,
+                                                            D is the feature dimension
+            y_vec: 1D or 2D array of datapoints, MxD, where M is the number of datapoints,
+                                                            D is the feature dimension
+
+        Returns:
+            2D matrix, NxM
+
+        Raises:
+            QiskitMachineLearningError:
+                A quantum instance or backend has not been provided
+            ValueError:
+                x_vec and/or y_vec are not one or two dimensional arrays
+                x_vec and y_vec have have incompatible dimensions
+                x_vec and/or y_vec have incompatible dimension with feature map and
+                    and feature map can not be modified to match.
+        """
+        raise NotImplementedError()
+
+    def _check_and_reshape(self, x_vec: np.ndarray, y_vec: np.ndarray = None) -> Tuple[np.ndarray]:
+        r"""
+        Performs checks on the dimensions of the input data x_vec and y_vec.
+        Reshapes the arrays so that `x_vec.shape = (N,D)` and `y_vec.shape = (M,D)`.
+        """
+        if not isinstance(x_vec, np.ndarray):
+            x_vec = np.asarray(x_vec)
+
+        if y_vec is not None and not isinstance(y_vec, np.ndarray):
+            y_vec = np.asarray(y_vec)
+
+        if x_vec.ndim > 2:
+            raise ValueError("x_vec must be a 1D or 2D array")
+
+        if x_vec.ndim == 1:
+            x_vec = x_vec.reshape(1, -1)
+
+        if y_vec is not None and y_vec.ndim > 2:
+            raise ValueError("y_vec must be a 1D or 2D array")
```

Comment:
One could avoid some duplication if checks for `x_vec` were extracted as a method. The same method would be calld on `y_vec` after checking it is not `None`.